File: src/lib/jira-client.ts
// Minimal Jira Cloud REST client for the CR-compliance checkpoint — reads
// only, via Basic auth (account email + API token), as an alternative to
// the Rovo MCP OAuth flow CP10 uses. Simpler for this checkpoint: a single
// static token, no interactive browser login / callback server required.
function jiraAuthHeader(): string {
  const email = process.env["JIRA_EMAIL"];
  const token = process.env["JIRA_API_TOKEN"];
  if (!email || !token) {
    throw new Error("JIRA_EMAIL / JIRA_API_TOKEN are not configured (set them in .env)");
  }
  return `Basic ${Buffer.from(`${email}:${token}`).toString("base64")}`;
}

function jiraBaseUrl(): string {
  const baseUrl = process.env["JIRA_BASE_URL"];
  if (!baseUrl) throw new Error("JIRA_BASE_URL is not configured (set it in .env)");
  return baseUrl.replace(/\/+$/, "");
}

// Atlassian Document Format node — comment bodies are a doc tree of these;
// only the shapes this module actually reads are typed.
type AdfNode = {
  type?: string;
  text?: string;
  attrs?: { text?: string };
  content?: AdfNode[];
};

export type JiraIssue = {
  key: string;
  fields: {
    summary?: string;
    status: {
      name: string;
      statusCategory: { key: string };
    };
    comment?: { comments: { body: AdfNode }[] };
  };
};

export async function getJiraIssue(issueKey: string): Promise<JiraIssue | null> {
  const res = await fetch(
    `${jiraBaseUrl()}/rest/api/3/issue/${encodeURIComponent(issueKey)}?fields=summary,status,comment`,
    { headers: { Authorization: jiraAuthHeader(), Accept: "application/json" } },
  );
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Jira issue lookup failed: HTTP ${res.status}`);
  return (await res.json()) as JiraIssue;
}

function adfPlainText(node: AdfNode): string {
  if (node.type === "text") return node.text ?? "";
  if (node.type === "mention") return node.attrs?.text ?? "";
  return (node.content ?? []).map(adfPlainText).join("");
}

function firstMention(node: AdfNode): string | null {
  if (node.type === "mention") return node.attrs?.text ?? null;
  for (const child of node.content ?? []) {
    const found = firstMention(child);
    if (found) return found;
  }
  return null;
}

// Looks for a comment like "The PR should be merged by @rahul" and returns
// the mentioned name ("rahul", "@" stripped). Deterministic text + mention
// match — never LLM-guessed — same approach as every other identity check
// in this app (CP10's sender match, CP38's authorized signer).
export function extractRequiredMerger(issue: JiraIssue): string | null {
  for (const comment of issue.fields.comment?.comments ?? []) {
    const text = adfPlainText(comment.body).toLowerCase();
    if (!text.includes("merged by")) continue;
    const mention = firstMention(comment.body);
    if (mention) return mention.replace(/^@/, "");
  }
  return null;
}

File: backend/app/clients/jira_client.py
# Minimal Jira Cloud REST client, Basic auth (account email + API token).
# Used by both CP10 (CAB approval status) and CR-compliance (Jira approval
# status) — CP10 no longer goes through Atlassian's Rovo MCP/OAuth flow.
from typing import Any

import httpx

from app.settings import settings


def _jira_auth() -> httpx.BasicAuth:
    email = settings.jira_email
    token = settings.jira_api_token
    if not email or not token:
        raise RuntimeError("JIRA_EMAIL / JIRA_API_TOKEN are not configured (set them in .env)")
    return httpx.BasicAuth(email, token)


def _jira_base_url() -> str:
    base_url = settings.jira_base_url
    if not base_url:
        raise RuntimeError("JIRA_BASE_URL is not configured (set it in .env)")
    return base_url.rstrip("/")


async def get_jira_issue(issue_key: str) -> dict[str, Any] | None:
    async with httpx.AsyncClient(auth=_jira_auth()) as client:
        res = await client.get(
            f"{_jira_base_url()}/rest/api/3/issue/{issue_key}",
            params={"fields": "summary,status,comment,reporter"},
            headers={"Accept": "application/json"},
        )
    if res.status_code == 404:
        return None
    if res.status_code >= 400:
        raise RuntimeError(f"Jira issue lookup failed: HTTP {res.status_code}")
    return res.json()


def _adf_plain_text(node: dict[str, Any]) -> str:
    if node.get("type") == "text":
        return node.get("text") or ""
    if node.get("type") == "mention":
        return (node.get("attrs") or {}).get("text") or ""
    return "".join(_adf_plain_text(child) for child in node.get("content") or [])


def _first_mention(node: dict[str, Any]) -> str | None:
    if node.get("type") == "mention":
        return (node.get("attrs") or {}).get("text")
    for child in node.get("content") or []:
        found = _first_mention(child)
        if found:
            return found
    return None


def extract_required_merger(issue: dict[str, Any]) -> str | None:
    """Looks for a comment like "The PR should be merged by @rahul" and
    returns the mentioned name ("rahul", "@" stripped). Deterministic text +
    mention match — never LLM-guessed."""
    comments = (issue.get("fields") or {}).get("comment", {}).get("comments") or []
    for comment in comments:
        body = comment.get("body") or {}
        text = _adf_plain_text(body).lower()
        if "merged by" not in text:
            continue
        mention = _first_mention(body)
        if mention:
            return mention.lstrip("@")
    return None
